Implement vulkan RHI

halx99 opened the pull request:
## Describe your changes

- vulkan RHI
- axslcc to 3.0.0
- enable glad vulkan support
- fix stencil manager verteices data
- refactor android jni and java surface class
- rename internal class name: `CommandBuffer` => `RenderContext`
- Add `tools/cmdline/shaderu.ps1` to upgrade exists shader to adapt vulkan RHI feature requirements


all shaders need migrate by `axmol/tools/cmdline/spirv-fix.ps`

## Target platforms

- Windows
- Linux

## Issue ticket number and link


## Checklist before requesting a review
### For each PR
- [ ] Add Copyright if it missed:   
      - `"Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."`
- [ ] I have performed a self-review of my code.
       
   Optional:
   - [ ] I have checked readme and add important infos to this PR.
   - [ ] I have added/adapted some tests too.
          
### For core/new feature PR
- [ ] I have checked readme and add important infos to this PR.
- [ ] I have added thorough tests.

### Axmol 3.x   ------------------------------------------------------------
### For each 3.x PR 
- [ ] Check the '#include "axmol.h"' and replace it with the needed headers.

halx99 committed:
WIP: Implement Vulkan RHI

halx99 commented:
/clang-format

aismann commented:
@halx99 
I belive vulkan have to have also an own build (win-vulkan) like:
win32-ogl, win32-d3d11, ...

halx99 commented:
> @halx99 I belive vulkan have to have also an own build (win-vulkan) like: win32-ogl, win32-d3d11, ...

Yes, but not current, because  there are still lots of work to do, current only beginFrame, endFrame, clearColor works

halx99 commented:
<img width="984" height="1213" alt="image" src="https://github.com/user-attachments/assets/5f64991e-6fb3-4585-b328-f4392ad980ef" />

paulocoutinhox commented:
Wow, nice. What is the advantage of have Vulkan support? Thanks.

halx99 commented:
/clang-format

halx99 committed:
Implement vk::RenderTextImpl::readPixels

halx99 committed:
Improve adapter selection

halx99 committed:
Improve VK resources management

halx99 committed:
Fix linux vulkan driver not support RGBA for surface

halx99 committed:
Fix descriptor sets assemble

halx99 committed:
vulkan: fix wayland create swapchain fail width/height is 0

halx99 committed:
vulkan: validate inframe state

halx99 committed:
Improve rhi::vk::RenderPipelineImpl

halx99 committed:
vulkan: fix primitive topology class mismatch

halx99 committed:
vulkan: fix generateMipmaps and add image layout tracker

halx99 commented:
So far, most 2D rendering works correctly (RenderTexture is known to be temporarily working). Mesh rendering is also functioning, except for instanced rendering. Next on the TODO list are terrain, grid, vulkan RHI optional descriptor set, and other...

halx99 committed:
vulkan: fix descriptor set layout sampler bindings

halx99 committed:
vulkan: captureScreen, captureNode now works

halx99 committed:
vulkan: mesh render instancing draw works

halx99 committed:
vulkan: skybox, terrain rendering works

halx99 committed:
Update axslcc to v3.0.0

reason:
Upgrading ensures reflection correctly handles cases like `sampler2D u_details[4]`,
where descriptorCount must match the array length rather than expanding into multiple bindings.

halx99 commented:
/clang-format

axmol-bot committed:
Committing clang-format changes

halx99 committed:
fixup

halx99 committed:
fix metal

halx99 committed:
vulkan: implement descriptor set cache

halx99 committed:
vulkan: optimized dynamicStateBits

halx99 committed:
vulkan: improve pipeline cache

halx99 committed:
renderer: fix shader test

halx99 committed:
vulkan: fix Grid draw by removing POT calc

- nextPOT caused incorrect UV coords in Grid vertex updates
- Axmol v3 has dropped GLES2.0, all RHIs now support NPOT textures
- effect tests verified working after change

halx99 commented:
Now, 99% cpp-tests works

halx99 committed:
fixup

halx99 commented:
/clang-format

axmol-bot committed:
Committing clang-format changes

halx99 committed:
vulkan: make android works

halx99 committed:
fixup

halx99 committed:
vulkan: process android VK_ERROR_SURFACE_LOST_KHR

halx99 commented:
/clang-format

axmol-bot committed:
Committing clang-format changes

aismann commented:
> > @halx99 I belive vulkan have to have also an own build (win-vulkan) like: win32-ogl, win32-d3d11, ...
> 
> Yes, but not current, because there are still lots of work to do, current only beginFrame, endFrame, clearColor works

Maybe time to add the vulkan build?
=> Easier for testing etc...

halx99 committed:
vulkan: add ci

aismann commented:
Please rename the artifact its has the same name like  `windows_d3d11_x64.zip`
Anyway: it seems vulkan is the fastest rendering pipeline (on windows) now.

halx99 committed:
Vulkan: use per-image acquire semaphores to fix Android flicker

- Changed acquire semaphores allocation to match swapchain image count
- Avoids reuse timing issues on certain Android GPU drivers
- Improves frame stability and reduces random flickering during acquire/present

aismann commented:
As far as Im can see:
Only the artifact section is done for VK (all others removed?)
<img width="377" height="44" alt="image" src="https://github.com/user-attachments/assets/e1acd885-4e2c-4072-a554-7ad675d6a442" />

At least D3D11 is missed
<img width="406" height="371" alt="image" src="https://github.com/user-attachments/assets/fda843c7-94a6-4f11-a45c-d7b52e54da4d" />

aismann commented:
`cpp-tests: autotest` (windows) runs without any error!
Great work!

aismann commented:
@halx99 
artefact: `windows_ogl_x64` is starting but closed fast (not working for me)

halx99 committed:
Improve code style

halx99 committed:
Include upload-artifact action in build workflow

Add artifact upload step for Windows D3D11 build

halx99 committed:
fixup

aismann commented:
@halx99 
`windows_ogl_x64` is still not starting (only the cmd is shown for few millisec)

halx99 committed:
fix ogl crash issue
